lib/ip_math.go

```diff
@@ -0,0 +1,65 @@
+package lib
+
+import (
+	"encoding/binary"
+	"math/big"
+	"net"
+)
+
+func ipAdd(input string, delta int) net.IP {
+	ip := net.ParseIP(input)
+	if ip.To4() != nil {
+		ipInt := ipToUint32(ip)
+		newIPInt := ipInt + uint32(delta)
+		newIP := uint32ToIP(newIPInt)
+		return newIP
+	} else {
+		ipInt := ipToBigInt(ip)
+		deltaBigInt := new(big.Int).SetInt64(int64(delta))
+		newIPInt := new(big.Int).Add(ipInt, deltaBigInt)
+		adjustedIPInt := adjustIPBigInt(newIPInt)
+		newIP := bigIntToIP(adjustedIPInt)
+		return newIP
+	}
+}
+
+func ipToUint32(ip net.IP) uint32 {
+	return binary.BigEndian.Uint32(ip.To4())
+}
+
+func uint32ToIP(ipInt uint32) net.IP {
+	ip := make(net.IP, net.IPv4len)
+	binary.BigEndian.PutUint32(ip, ipInt)
+	return ip
+}
+
+func ipToBigInt(ip net.IP) *big.Int {
+	ipInt := new(big.Int)
+	ipInt.SetBytes(ip)
+	return ipInt
+}
+
+func bigIntToIP(ipInt *big.Int) net.IP {
```

Comment:
The algorithm for this can be simplified now to simply truncate the extra bytes of the BigInt. Simply, `copy(ip[:], ipIntBytes[:net.IPv6len]` (I think! - need to check if this crashes or not on lower ipv6 values)

Reply:
This gives runtime error: slice bounds out of range. I have rather rewritten the code with an extra layer of checking in more simplified way.  The code first calculates the length to copy (copyLength) as the length of 'ipIntBytes'. If 'copyLength' exceeds 'net.IPv6len', it is truncated to 'net.IPv6len'. Then, only the last 'copyLength' bytes from 'ipIntBytes' are copied to the end of the ip slice.